articles/ml_system_design/real_time_news_search_using_upstash_kafka_and_vector/upstash_ingest/upstash_ingest/models.py

```diff
@@ -0,0 +1,225 @@
+        }
+        refined.full_text = ".".join([common.title, common.description])
+        return refined
+
+
+class ChunkedDocument(BaseModel):
+    doc_id: str
+    chunk_id: str
+    full_raw_text: str
+    text: str
+    metadata: Dict[str, Union[str, Any]]
+
+    @classmethod
+    def from_refined(
+        cls, refined_doc: RefinedDocument, embedding_model: TextEmbedder
+    ) -> list["ChunkedDocument"]:
+        chunks = ChunkedDocument.chunkenize(refined_doc.full_text, embedding_model)
+
+        return [
+            cls(
+                doc_id=refined_doc.doc_id,
+                chunk_id=hashlib.md5(chunk.encode()).hexdigest(),
+                full_raw_text=refined_doc.full_text,
+                text=chunk,
+                metadata=refined_doc.metadata,
+            )
+            for chunk in chunks
+        ]
+
+    @staticmethod
+    def chunkenize(text: str, embedding_model: TextEmbedder) -> list[str]:
+        text_sections = RECURSIVE_SPLITTER.split_text(text=text)
+        chunks = []
+        for text_section in text_sections:
+            chunks.extend(
+                chunk_by_attention_window(text_section, embedding_model.tokenizer)
+            )
+
+        return chunks
+
+
+class EmbeddedDocument(BaseModel):
+    doc_id: str
+    chunk_id: str
+    full_raw_text: str
+    text: str
+    embeddings: List[float]
+    metadata: Dict[str, Union[str, Any]] = {}
+
+    @classmethod
+    def from_chunked(
+        cls, chunked_doc: ChunkedDocument, embedding_model: TextEmbedder
+    ) -> "EmbeddedDocument":
+        return cls(
+            doc_id=chunked_doc.doc_id,
+            chunk_id=chunked_doc.chunk_id,
+            full_raw_text=chunked_doc.full_raw_text,
+            text=chunked_doc.text,
+            embeddings=embedding_model(chunked_doc.text, to_list=True),
+            metadata=chunked_doc.metadata,
+        )
+
+    def to_payload(self) -> tuple[str, List[float], dict]:
+        return (self.chunk_id, self.embeddings, self.metadata)
+
+    def __repr__(self) -> str:
+        return f"EmbeddedDocument(doc_id={self.doc_id}, chunk_id={self.chunk_id})"
+
+
+class CommonDocument(BaseModel):
+    article_id: str = Field(default_factory=lambda: str(uuid4()))
+    title: str
+    url: str = Field(default="")
+    published_at: str = Field(
+        default_factory=lambda: datetime.now().strftime("%Y-%m-%d %H:%M:%S")
+    )
+    source_name: str = Field(default="")
+    image_url: Optional[str] = Field(default="N/A")
+    author: Optional[str] = Field(default="Unknown")
+    description: Optional[str] = Field(default="")
+    content: Optional[str] = Field(default="")
+
+    @field_validator("title", "description", "content")
+    def clean_text_fields(cls, v):
+        if v is None:
+            return ""
+        return clean_text(v)
+
+    @field_validator("url", "image_url")
+    def clean_url_fields(cls, v):
+        if v is None:
+            return ""
+        v = remove_html_tags(v)
```

Comment:
Personally, I would move the cleaning steps to the factory method, as you should perform the validation here, not the transformation, per se. But this is just my preference.

Reply:
Your suggestion is to define field clean steps in pydantic Field(default-factory) ?

Reply:
The Field(...default_factory) assigns a def value to the field if None or invalid. Won't work with methods that require arguments because the class fields will instantiate before the constructor.

From pydantic docs: `@field_validator("...")` is used to validate (change, format) fields